fhirpath/src/main/java/au/csiro/pathling/fhirpath/encoding/QuantityEncoding.java

```diff
@@ -302,28 +247,84 @@ public static Column encodeLiteral(@Nonnull final FhirPathQuantity quantity) {
-        // scale.
-        numericColumn.cast(DecimalCustomCoder.decimalType()),
-        // We cannot encode the scale of the results of arithmetic operations.
-        lit(null),
-        lit(null),
-        lit("1"),
-        lit(FhirPathQuantity.UCUM_SYSTEM_URI),
-        lit("1"),
-        // we do not need to normalize this as the unit is always "1"
-        // so it will be comparable with other quantities with unit "1"
-        lit(null),
-        lit(null),
-        lit(null));
+    // Cast value to decimal type
+    final Column decimalValue = numericColumn.cast(DecimalCustomCoder.decimalType());
+
+    // Return fully null struct when value is null to maintain FHIRPath empty collection semantics
+    return when(decimalValue.isNotNull(),
+        toStruct(
+            lit(null),
+            decimalValue,
+            // We cannot encode the scale of the results of arithmetic operations.
+            lit(null),
+            lit(null),
+            lit("1"),
+            lit(UcumUnit.UCUM_SYSTEM_URI),
+            lit("1"),
+            // we do not need to normalize this as the unit is always "1"
+            // so it will be comparable with other quantities with unit "1"
+            lit(null),
+            lit(null),
+            lit(null)))
+        .otherwise(lit(null).cast(dataType()));
+  }
+
+  /**
+   * Encodes a FhirPathQuantity to a Row representation.
+   * <p>
+   * This method converts a FhirPathQuantity into the Row representation used by Spark. It handles
+   * canonicalization for both UCUM units and calendar duration units.
+   *
+   * @param quantity the FhirPathQuantity to encode
+   * @return the Row representation of the quantity
+   */
+  @Nonnull
+  public static Row encode(@Nonnull final FhirPathQuantity quantity) {
+    final BigDecimal value = quantity.getValue();
+    final BigDecimal canonicalizedValue;
+    final String canonicalizedCode;
+
+    if (quantity.isUcum()) {
+      // If it is a UCUM Quantity, use the UCUM library to canonicalize the value and code.
+      canonicalizedValue = Ucum.getCanonicalValue(value, quantity.getCode());
+      canonicalizedCode = Ucum.getCanonicalCode(value, quantity.getCode());
+    } else if (quantity.isCalendarDuration() &&
+        CALENDAR_DURATION_TO_UCUM.containsKey(quantity.getCode())) {
+      // If it is a (supported) calendar duration, get the corresponding UCUM unit and then use
+      // the UCUM library to canonicalize the value and code.
+      final String resolvedCode = CALENDAR_DURATION_TO_UCUM.get(quantity.getCode());
+      canonicalizedValue = Ucum.getCanonicalValue(value, resolvedCode);
+      canonicalizedCode = Ucum.getCanonicalCode(value, resolvedCode);
+    } else {
+      // If it is neither a UCUM Quantity nor a calendar duration, it will not have a
+      // canonicalized form available.
+      canonicalizedValue = null;
+      canonicalizedCode = null;
+    }
```

Comment:
This code seems to be duplicated from `encodeLiteral` - perhaps we can extract it.